File: src/lib/ahanu/geo.ts
import { METERS_PER_NM, NM_PER_DEG_LAT } from "./constants";
import type { LatLon } from "./types";

const R_NM = 3440.065; // Earth radius in nautical miles

export function toRad(d: number) {
  return (d * Math.PI) / 180;
}
export function toDeg(r: number) {
  return (r * 180) / Math.PI;
}

export function haversineNm(a: LatLon, b: LatLon): number {
  const dLat = toRad(b.lat - a.lat);
  const dLon = toRad(b.lon - a.lon);
  const s =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * Math.sin(dLon / 2) ** 2;
  return 2 * R_NM * Math.asin(Math.min(1, Math.sqrt(s)));
}

export function initialBearing(a: LatLon, b: LatLon): number {
  const y = Math.sin(toRad(b.lon - a.lon)) * Math.cos(toRad(b.lat));
  const x =
    Math.cos(toRad(a.lat)) * Math.sin(toRad(b.lat)) -
    Math.sin(toRad(a.lat)) * Math.cos(toRad(b.lat)) * Math.cos(toRad(b.lon - a.lon));
  return (toDeg(Math.atan2(y, x)) + 360) % 360;
}

export function destination(start: LatLon, bearingDeg: number, nm: number): LatLon {
  const dR = nm / R_NM;
  const br = toRad(bearingDeg);
  const lat1 = toRad(start.lat);
  const lon1 = toRad(start.lon);
  const lat2 = Math.asin(
    Math.sin(lat1) * Math.cos(dR) + Math.cos(lat1) * Math.sin(dR) * Math.cos(br),
  );
  const lon2 =
    lon1 +
    Math.atan2(
      Math.sin(br) * Math.sin(dR) * Math.cos(lat1),
      Math.cos(dR) - Math.sin(lat1) * Math.sin(lat2),
    );
  return { lat: toDeg(lat2), lon: ((toDeg(lon2) + 540) % 360) - 180 };
}

export function alongTrack(a: LatLon, b: LatLon, t: number): LatLon {
  const nm = haversineNm(a, b);
  if (nm < 1e-6) return a;
  return destination(a, initialBearing(a, b), nm * t);
}

export function pathLengthNm(pts: LatLon[]): number {
  let n = 0;
  for (let i = 1; i < pts.length; i++) n += haversineNm(pts[i - 1]!, pts[i]!);
  return n;
}

export function interpolatePath(pts: LatLon[], t: number): { pos: LatLon; cog: number } {
  if (pts.length === 0) return { pos: { lat: 0, lon: 0 }, cog: 0 };
  if (pts.length === 1) return { pos: pts[0]!, cog: 0 };
  const total = pathLengthNm(pts);
  let remain = Math.max(0, Math.min(1, t)) * total;
  for (let i = 1; i < pts.length; i++) {
    const a = pts[i - 1]!;
    const b = pts[i]!;
    const seg = haversineNm(a, b);
    if (remain <= seg || i === pts.length - 1) {
      const u = seg < 1e-6 ? 0 : remain / seg;
      return { pos: alongTrack(a, b, u), cog: initialBearing(a, b) };
    }
    remain -= seg;
  }
  const last = pts[pts.length - 1]!;
  return { pos: last, cog: initialBearing(pts[pts.length - 2]!, last) };
}

export function nmToLatLonDelta(lat: number, dNmNorth: number, dNmEast: number): LatLon {
  return {
    lat: dNmNorth / NM_PER_DEG_LAT,
    lon: dNmEast / (NM_PER_DEG_LAT * Math.cos(toRad(lat))),
  };
}

export function formatLat(lat: number): string {
  const hem = lat >= 0 ? "N" : "S";
  const a = Math.abs(lat);
  const d = Math.floor(a);
  const m = (a - d) * 60;
  return `${d}°${m.toFixed(3).padStart(6, "0")}'${hem}`;
}

export function formatLon(lon: number): string {
  const hem = lon >= 0 ? "E" : "W";
  const a = Math.abs(lon);
  const d = Math.floor(a);
  const m = (a - d) * 60;
  return `${d}°${m.toFixed(3).padStart(6, "0")}'${hem}`;
}

export function formatCoord(p: LatLon): string {
  return `${formatLat(p.lat)}  ${formatLon(p.lon)}`;
}

export function compass(deg: number): string {
  const dirs = ["N", "NNE", "NE", "ENE", "E", "ESE", "SE", "SSE", "S", "SSW", "SW", "WSW", "W", "WNW", "NW", "NNW"];
  return dirs[Math.round(((deg % 360) + 360) % 360 / 22.5) % 16]!;
}

export function metersToFathoms(m: number) {
  return m / 1.8288;
}

export function metersToFeet(m: number) {
  return m * 3.28084;
}

export function hoursToHm(h: number): string {
  const s = Math.max(0, h);
  const hh = Math.floor(s);
  const mm = Math.round((s - hh) * 60);
  return `${hh}h ${mm.toString().padStart(2, "0")}m`;
}

export function lonLatToXY(p: LatLon, origin: LatLon): { x: number; y: number } {
  const dLat = p.lat - origin.lat;
  const dLon = p.lon - origin.lon;
  return {
    x: dLon * NM_PER_DEG_LAT * Math.cos(toRad(origin.lat)) * METERS_PER_NM,
    y: dLat * NM_PER_DEG_LAT * METERS_PER_NM,
  };
}

/** Closed ring as lon/lat pairs, clockwise from north. */
export function circlePoints(center: LatLon, radiusNm: number, n = 64): LatLon[] {
  const pts: LatLon[] = [];
  const steps = Math.max(8, n);
  for (let i = 0; i < steps; i++) {
    pts.push(destination(center, (i / steps) * 360, radiusNm));
  }
  pts.push(pts[0]!);
  return pts;
}

export function circleRingGeo(
  center: LatLon,
  radiusNm: number,
  n = 64,
): GeoJSON.FeatureCollection {
  const ring = circlePoints(center, radiusNm, n).map((p) => [p.lon, p.lat] as [number, number]);
  return {
    type: "FeatureCollection",
    features: [
      {
        type: "Feature",
        properties: { radiusNm },
        geometry: { type: "LineString", coordinates: ring },
      },
    ],
  };
}

export function measureSummary(pts: LatLon[]): { nm: number; bearing: number; legs: number } {
  if (pts.length < 2) return { nm: 0, bearing: 0, legs: 0 };
  return {
    nm: pathLengthNm(pts),
    bearing: initialBearing(pts[pts.length - 2]!, pts[pts.length - 1]!),
    legs: pts.length - 1,
  };
}

File: flutter/lib/src/geo.dart
/// Nautical geometry. Same formulas as `src/lib/ahanu/geo.ts`.
///
/// Earth radius is **3440.065 nautical miles**. Distances returned by
/// [haversineNm] are nautical miles. Depth helpers convert metres.
library;

import 'dart:math' as math;

import 'constants.dart';
import 'types.dart';

/// Earth radius in nautical miles. Matches `R_NM` in `geo.ts`.
const earthRadiusNm = 3440.065;

double toRad(double d) => (d * math.pi) / 180;
double toDeg(double r) => (r * 180) / math.pi;

double haversineNm(LatLon a, LatLon b) {
  final dLat = toRad(b.lat - a.lat);
  final dLon = toRad(b.lon - a.lon);
  final s = math.sin(dLat / 2) * math.sin(dLat / 2) +
      math.cos(toRad(a.lat)) *
          math.cos(toRad(b.lat)) *
          math.sin(dLon / 2) *
          math.sin(dLon / 2);
  return 2 * earthRadiusNm * math.asin(math.min(1.0, math.sqrt(s)));
}

double initialBearing(LatLon a, LatLon b) {
  final y = math.sin(toRad(b.lon - a.lon)) * math.cos(toRad(b.lat));
  final x = math.cos(toRad(a.lat)) * math.sin(toRad(b.lat)) -
      math.sin(toRad(a.lat)) *
          math.cos(toRad(b.lat)) *
          math.cos(toRad(b.lon - a.lon));
  return (toDeg(math.atan2(y, x)) + 360) % 360;
}

LatLon destination(LatLon start, double bearingDeg, double nm) {
  final dR = nm / earthRadiusNm;
  final br = toRad(bearingDeg);
  final lat1 = toRad(start.lat);
  final lon1 = toRad(start.lon);
  final lat2 = math.asin(
    math.sin(lat1) * math.cos(dR) +
        math.cos(lat1) * math.sin(dR) * math.cos(br),
  );
  final lon2 = lon1 +
      math.atan2(
        math.sin(br) * math.sin(dR) * math.cos(lat1),
        math.cos(dR) - math.sin(lat1) * math.sin(lat2),
      );
  return LatLon(
    lat: toDeg(lat2),
    lon: ((toDeg(lon2) + 540) % 360) - 180,
  );
}

LatLon alongTrack(LatLon a, LatLon b, double t) {
  final nm = haversineNm(a, b);
  if (nm < 1e-6) return a;
  return destination(a, initialBearing(a, b), nm * t);
}

double pathLengthNm(List<LatLon> pts) {
  var n = 0.0;
  for (var i = 1; i < pts.length; i++) {
    n += haversineNm(pts[i - 1], pts[i]);
  }
  return n;
}

class PathSample {
  const PathSample({required this.pos, required this.cog});

  final LatLon pos;
  final double cog;
}

PathSample interpolatePath(List<LatLon> pts, double t) {
  if (pts.isEmpty) {
    return const PathSample(pos: LatLon(lat: 0, lon: 0), cog: 0);
  }
  if (pts.length == 1) return PathSample(pos: pts.first, cog: 0);
  final total = pathLengthNm(pts);
  var remain = t.clamp(0.0, 1.0) * total;
  for (var i = 1; i < pts.length; i++) {
    final a = pts[i - 1];
    final b = pts[i];
    final seg = haversineNm(a, b);
    if (remain <= seg || i == pts.length - 1) {
      final u = seg < 1e-6 ? 0.0 : remain / seg;
      return PathSample(pos: alongTrack(a, b, u), cog: initialBearing(a, b));
    }
    remain -= seg;
  }
  final last = pts.last;
  return PathSample(
    pos: last,
    cog: initialBearing(pts[pts.length - 2], last),
  );
}

LatLon nmToLatLonDelta(double lat, double dNmNorth, double dNmEast) {
  return LatLon(
    lat: dNmNorth / NM_PER_DEG_LAT,
    lon: dNmEast / (NM_PER_DEG_LAT * math.cos(toRad(lat))),
  );
}

String formatLat(double lat) {
  final hem = lat >= 0 ? 'N' : 'S';
  final a = lat.abs();
  final d = a.floor();
  final m = (a - d) * 60;
  return "$d°${m.toStringAsFixed(3).padLeft(6, '0')}'$hem";
}

String formatLon(double lon) {
  final hem = lon >= 0 ? 'E' : 'W';
  final a = lon.abs();
  final d = a.floor();
  final m = (a - d) * 60;
  return "$d°${m.toStringAsFixed(3).padLeft(6, '0')}'$hem";
}

String formatCoord(LatLon p) => '${formatLat(p.lat)}  ${formatLon(p.lon)}';

const _compassDirs = [
  'N',
  'NNE',
  'NE',
  'ENE',
  'E',
  'ESE',
  'SE',
  'SSE',
  'S',
  'SSW',
  'SW',
  'WSW',
  'W',
  'WNW',
  'NW',
  'NNW',
];

String compass(double deg) {
  final idx = ((((deg % 360) + 360) % 360) / 22.5).round() % 16;
  return _compassDirs[idx];
}

double metersToFathoms(double m) => m / 1.8288;

double metersToFeet(double m) => m * 3.28084;

String hoursToHm(double h) {
  final s = math.max(0.0, h);
  final hh = s.floor();
  final mm = ((s - hh) * 60).round();
  return '${hh}h ${mm.toString().padLeft(2, '0')}m';
}

class MetresXY {
  const MetresXY({required this.x, required this.y});

  final double x;
  final double y;
}

MetresXY lonLatToXY(LatLon p, LatLon origin) {
  final dLat = p.lat - origin.lat;
  final dLon = p.lon - origin.lon;
  return MetresXY(
    x: dLon * NM_PER_DEG_LAT * math.cos(toRad(origin.lat)) * METERS_PER_NM,
    y: dLat * NM_PER_DEG_LAT * METERS_PER_NM,
  );
}
